Toolchain/Patches/llvm.patch

```diff
@@ -0,0 +1,923 @@
++} // end namespace tools
++
++namespace toolchains {
++
++class LLVM_LIBRARY_VISIBILITY Serenity : public Generic_ELF {
++public:
++  Serenity(const Driver &D, const llvm::Triple &Triple,
++        const llvm::opt::ArgList &Args);
++
++  void AddClangSystemIncludeArgs(const llvm::opt::ArgList &DriverArgs,
++        llvm::opt::ArgStringList &CC1Args) const override;
++
++  void AddClangCXXStdlibIncludeArgs(const llvm::opt::ArgList &DriverArgs,
++        llvm::opt::ArgStringList &CC1Args) const override;
++
++
++  RuntimeLibType GetDefaultRuntimeLibType() const override {
++    return ToolChain::RLT_CompilerRT;
++  }
++
++  UnwindLibType GetDefaultUnwindLibType() const override {
++    return ToolChain::UNW_CompilerRT;
++  }
++
++  CXXStdlibType GetDefaultCXXStdlibType() const override {
++    return ToolChain::CST_Libcxx;
++  }
++  const char *getDefaultLinker() const override {
++    return "ld.lld";
++  }
++
++  bool HasNativeLLVMSupport() const override { return true; }
++  bool IsIntegratedAssemblerDefault() const override { return true; }
++
++  unsigned GetDefaultDwarfVersion() const override { return 5; }
++protected:
++  Tool *buildLinker() const override;
++};
++
++} // end namespace toolchains
++} // end namespace driver
++} // end namespace clang
++
++#endif // LLVM_CLANG_LIB_DRIVER_TOOLCHAINS_SERENITY_H
+diff --git a/clang/lib/Frontend/InitHeaderSearch.cpp b/clang/lib/Frontend/InitHeaderSearch.cpp
+index bc31445d6..a4a3fbc89 100644
+--- a/clang/lib/Frontend/InitHeaderSearch.cpp
++++ b/clang/lib/Frontend/InitHeaderSearch.cpp
+@@ -229,6 +229,7 @@ void InitHeaderSearch::AddDefaultCIncludePaths(const llvm::Triple &triple,
+     case llvm::Triple::PS4:
+     case llvm::Triple::ELFIAMCU:
+     case llvm::Triple::Fuchsia:
++    case llvm::Triple::Serenity:
+       break;
+     case llvm::Triple::Win32:
+       if (triple.getEnvironment() != llvm::Triple::Cygnus)
+@@ -428,6 +429,7 @@ void InitHeaderSearch::AddDefaultIncludePaths(const LangOptions &Lang,
+   case llvm::Triple::Solaris:
+   case llvm::Triple::WASI:
+   case llvm::Triple::AIX:
++  case llvm::Triple::Serenity:
+     return;
+ 
+   case llvm::Triple::Win32:
+diff --git a/libcxx/include/__config b/libcxx/include/__config
+index a3838c89e..8a20731ef 100644
+--- a/libcxx/include/__config
++++ b/libcxx/include/__config
+@@ -1132,6 +1132,7 @@ extern "C" _LIBCPP_FUNC_VIS void __sanitizer_annotate_contiguous_container(
+       defined(__CloudABI__) || \
+       defined(__sun__) || \
+       defined(__MVS__) || \
++      defined(__serenity__) || \
+       (defined(__MINGW32__) && __has_include(<pthread.h>))
+ #    define _LIBCPP_HAS_THREAD_API_PTHREAD
+ #  elif defined(__Fuchsia__)
+diff --git a/libcxx/include/__string b/libcxx/include/__string
+index d8b672e4c..638cf1698 100644
+--- a/libcxx/include/__string
++++ b/libcxx/include/__string
+@@ -451,6 +451,7 @@ char_traits<char>::find(const char_type* __s, size_t __n, const char_type& __a)
+ }
+ 
+ 
++#ifndef __serenity__
```

Comment:
Why do you need to patch up the libcxx/ headers if libcxx isn't part of LLVM_ENABLE_PROJECT?

Reply:
`libcxx` is built in a separate batch, because I didn't manage to cram all those config options into the `llvm+clang` build.